src/init.cpp

```diff
@@ -1777,6 +1777,7 @@ bool AppInitMain(InitInterfaces& interfaces)
     connOptions.nLocalServices = nLocalServices;
     connOptions.nMaxConnections = nMaxConnections;
     connOptions.nMaxOutbound = std::min(MAX_OUTBOUND_CONNECTIONS, connOptions.nMaxConnections);
+    connOptions.nMaxOutboundBlocksOnly = std::min(MAX_BLOCKS_ONLY_CONNECTIONS, std::max(connOptions.nMaxOutbound-6, 0));
```

Comment:
This `- 6` is a little confusing and requires that you look up the default value of `connOptions.nMaxOutbound` to understand what this is doing. Also echoing @dongcarl comments that this doesn't work well when `connOptions.nMaxOutbound <= 6`.


Given that this fails for particular values and you only want 2 block only outbound why not change this to:

```
connOptions.nMaxOutboundBlocksOnly = MAX_BLOCKS_ONLY_CONNECTIONS
```

Reply:
Perhaps the accounting logic should be improved to be clearer, but I believe the behavior here is at least (roughly) correct. The way this works is that if a user sets `-maxconnections` very low, that will be a hard limit on the number of inbound, outbound and outbound-blocks-only connections.  Before this PR, it would take the user setting their max connections value to less than 8 to change the number of outbound connections we make (we first reduce all the inbound we take and only then reduce the outbound).

This PR adds new blocks-only connections and preferences those connections over regular outbound connections (whenever we make an outbound connection, we check to see if we are below our blocks-only target, and if so we initiate the connection as blocks-only).  As a result, I thought it didn't make sense to have any blocks-only links if we have too few regular outbound connections, because I don't think it makes sense to have blocks-only links at the expense of transaction relay.

In particular, the problem with the suggestion of just setting nMaxOutboundBlocksOnly to 2 is that if the user passes `-maxconnections=2`, then they'll only have 2 blocksonly peers and won't receive any transactions.

So I believe the way I implemented things here is that we try to reserve 6 connections as regular outbound, and then if we add additional connections we will have those be blocks-only until we get to 2, and then any further connections are regular outbounds up to 8, and then finally any additional ones are used for feelers and inbounds.

Of course we could change this to be more conservative and only allocate blocks-only connections if we have all 8 regular outbounds.  However I don't think we should be less conservative and risk having fewer regular outbounds.

Reply:
I think just do std::min(MAX_BLOCKS_ONLY_CONNECTIONS, std::max(connOptions.nMaxConnections - MAX_OUTBOUND_CONNECTIONS, 0)).

Reply:
@TheBlueMatt  I pushed a commit that changes the behavior so that the first 8 outbounds are reserved to be regular full-relay peers, and only then do we allocate up to 2 blocksonly peers.